text/3502-cargo-script.md

```diff
@@ -0,0 +1,1205 @@
+- Feature Name: `cargo-script`
+- Start Date: 2023-03-31
+- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
+- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
+- RFC PR: [rust-lang/rfcs#3502](https://github.com/rust-lang/rfcs/pull/3502)
+- Rust Issue: [rust-lang/cargo#12207](https://github.com/rust-lang/cargo/issues/12207)
+
+# Summary
+[summary]: #summary
+
+This RFC adds support for single-file bin packages in cargo.
+Single-file bin packages are rust source files with an embedded manifest and a
+`main`.
+These files will be accepted by cargo commands as `--manifest-path` just like `Cargo.toml` files.
+`cargo` will be modified to accept `cargo <file>.rs` as a shortcut to `cargo
+run --manifest-path <file>.rs`;
+this allows placing `cargo` in a `#!` line for directly running these files.
+
+Support for single-file lib packages, publishing, and workspace support is
+deferred out.
+
+Example:
+````rust
+#!/usr/bin/env cargo
+```cargo
```

Comment:
`cargo` has been specified in the shebang above already. Why does the user need to specify it twice? It could just allow a naked ``` fence instead (at least if a shebang is present). We know that we are in a cargo-script file in that instance, so we have the freedom to set a "default" case for what to do with fences that have no specifier.

IMO defaulting them to the manifest does not paint us in a corner while saving us character overhead.

Reply:
Specifying it leaves us much more room for future expansion, such as embedding other associated files or serving as frontmatter for other purposes. And it will make the job of syntax highlighters much easier.

If we want less appearance of redundancy, at the expense of more characters, we could use `manifest` or similar. But that doesn't seem worth it.

Reply:
The start choice tends to stick around. In response to my suggestion in https://github.com/rust-lang/cargo/pull/12301 , I have concerns that it's an unfamiliar way of specifying the edition. compare it to the PR from when the edition mechanism was introduced, there one of the comments against was for a similar reason to the one you reject `cargo` defaulting now: that we don't know if we want to make `2018-pre` a thing or not. Not that I agree with them but a similar effect might occur if we require the cargo specifier to be present.

Regarding whether it should be discussed in #3503 or not, IMO that RFC should not be as restricted to precisely what cargo needs, but be more generic. Ideally when we add support for non manifest embeddings, we shouldn't have to make another lang RFC but handle it wholly inside cargo-script.

Reply:
> Ideally when we add support for non manifest embeddings, we shouldn't have to make another lang RFC

i doubt we'd need an RFC; likely a change proposal would do.

Reply:
#3503 was merged which has us make the infostring optional